### ISSUE TITLE
When launching in a multi-GPU / processing setup, not all instances are observed

### ISSUE THREAD

**Issue by gorjanradevski** [huggingface/accelerate#2676]
### System Info

```Shell
- `Accelerate` version: 0.28.0
- Platform: Linux-4.18.0-477.43.1.el8_8.x86_64-x86_64-with-glibc2.28
- Python version: 3.10.14
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.2.1 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 503.19 GB
- GPU type: NVIDIA A100-SXM4-80GB
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: bf16
	- use_cpu: False
	- debug: False
	- num_processes: 2
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
	- tpu_env: []
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I have the following script (`test_accelerate.py`):

```python
from accelerate import Accelerator
import torch
from tqdm import tqdm
from collections import defaultdict


class TestDataset(torch.utils.data.Dataset):

    def __init__(self, start, end):
        self.samples = list(range(start, end))

    def __len__(self):
        return len(self.samples)

    def __getitem__(self, idx):
        return {"id": self.samples[idx]}


class Collater:

    def __init__(self):
        pass

    def __call__(self, batch):
        output = defaultdict(list)
        # Pre-pack
        for b in batch:
            for k in b.keys():
                output[k].append(b[k])

        output["id"] = torch.tensor(output["id"])

        return output


@torch.no_grad()
def main():
    accelerator = Accelerator()
    dataset = TestDataset(start=0, end=100_000)
    train_loader = torch.utils.data.DataLoader(
        dataset,
        batch_size=2,
        num_workers=12,
        shuffle=True,
        pin_memory=True,
        collate_fn=Collater(),
    )
    train_loader = accelerator.prepare(train_loader)
    for epoch in range(10):
        unique_samples = set()
        with tqdm(
            total=len(train_loader), disable=not accelerator.is_main_process
        ) as pbar:
            for batch in train_loader:
                # Count uniques
                gathered_ids = accelerator.gather(batch["id"])
                gathered_ids = gathered_ids.cpu().tolist()
                unique_samples.update(set(gathered_ids))
                # Update progress bar
                pbar.update(1)
        if accelerator.is_main_process:
            print(f"Observed: {len(unique_samples)} samples...")


if __name__ == "__main__":
    main()
```

Which I'm launching as: `accelerate launch --num_processes 4 test_accelerate.py`.

### Expected behavior

When launching the script not all samples from the dataset are observed in each "epoch".  However, if I set `shuffle=False` in the data loader (i.e., like I would do during inference), then all samples are observed. On the other hand, I also assumed that when training `shuffle=True`, that the model will be exposed to all samples from the dataset in each epoch.

Furthermore, if I launch the script as `accelerate launch --num_processes 1 test_accelerate.py`, all samples from the "dataset" are observed in each "epoch".

**Comment by gorjanradevski** [comment]
I also tested the same script on a different machine, with the latest accelerate version, but the result was the same when I launch as `accelerate launch --num_processes 4 test_accelerate.py`:
```
- `Accelerate` version: 0.29.2
- Platform: Linux-5.4.0-45-generic-x86_64-with-glibc2.31
- `accelerate` bash location: /home/gradevski/miniconda3/envs/accelerate/bin/accelerate
- Python version: 3.10.14
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.2.2 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- System RAM: 251.57 GB
- GPU type: NVIDIA GeForce RTX 2080 Ti
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: MULTI_GPU
	- mixed_precision: no
	- use_cpu: False
	- debug: False
	- num_processes: 4
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- enable_cpu_affinity: False
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
	- tpu_env: []
```

@muellerzr any idea what might be going on?

**Comment by muellerzr** [comment]
> On the other hand, I also assumed that when training shuffle=True, that the model will be exposed to all samples from the dataset in each epoch.

It should still shard the dataset across, so e.g. samples 0-3 go to GPU 0, 4-7 to GPU 1, and then if `drop_last` is enabled if any part of it is partial then we cut it off to the nearest completion.

Will look into this asap

**Comment by muellerzr** [comment]
How many samples are reported? I don't think you gave us the original "Observed" count

**Comment by gorjanradevski** [comment]
Out of `100_000` samples around `68_000` are reported. I just tested all released `accelerate` versions, the first version that reports the correct number is `0.24.1`.

**Comment by muellerzr** [comment]
Your phrasing is a bit confusing there to me. Are we saying:

- After (and including) 0.24.1 the values reported are incorrect, the 68_000
- Prior to 0.24.1 the values reported are incorrect, the 68_000

**Comment by gorjanradevski** [comment]
- From `0.29.2` until `0.25.0` the incorrect number is reported.
- First version that reports the correct number is `0.24.1`.
- I didn't test the older ones.

**Comment by muellerzr** [comment]
Thanks! I'll look into this

**Comment by gorjanradevski** [comment]
Let me know, I'm curious what is going on 😄. Besides, thanks a lot for all of the work on this library, it has been truly a pleasure using it 😄.

**Comment by gorjanradevski** [comment]
Hey @muellerzr, did you have time to look into this?
